Towards fullgeneric 4

cseed opened the pull request:
Spicy meatball for you, @tpoterba.  I hope I didn't step on your feet too much.

read{_table} and write now read and write rows, stolen and tweaked up from tpoterba/unsafe-rowstore-2.  (Nice work!)  Parquet is gone.  Not having to scan all the partitions feels so nice, and I'm just working on tiny examples on my laptop.

Added RegionValueBuilder which is useful for ... building region-based values (values allocated in a MemoryBuffer).  `import_vcf` uses it to produces `RegionValues`/`UnsafeRow`.  I left in `UnsafeRowBuilder`, but it is not being used (except by the tests).  We should port over the region => region optimization, and remove it.  I feel like this could be used to write our own non-consing Parquet importer easily (that supports nested fields!)  Also, our own VCF parser is now trivial to drop in, esp. for genotypes.

Added UnsafeIndexedSeqAnnotation and pulling native complex types out of unsafe rows.

Cleaned up read/writing VDS/KT metadata files.

Got rid of `RowGenotype`, wrote `buildGenotypeExtractor` to be much better.

I handled the serialization issue a slightly different way.  See `BroadcastTypeTree`.  Including your Kryo optimizations from unsafe-rowstore-2 would be good, too.

It is still not as fast as 0.1, but generic and getting closer.  This change has a lot of upside.  Making things mutable now is trivial (just remove to `region.copy()` in `LoadVCF`, `ReadRowsRDD`, etc.)  Tests spend a lot of time in methods that should eventually go away (e.g. `UnsafeRow.read`).  The main problem is that the rowstore with naive encodings is about 4x larger (compressed) than the corresponding 0.1 VDS (profile225, 2.0GB => 7.8GB) and a huge amount of time is spent in LZ4 compression.  I have a plan for this.

tpoterba commented:
I love it! This will probably take a couple days to review.

tpoterba commented:
I'll digest over the next day or so, then let's sit down and go through some of the design choices + trajectory together.

jbloom22 commented:
This is exciting! Could we discuss the reasoning for some of these changes at check-in?

tpoterba commented on src/main/scala/is/hail/annotations/ReadRowsRDD.scala:
```diff
@@ -0,0 +1,134 @@
+
+case class ReadRowsRDDPartition(index: Int) extends Partition
+
+class ReadRowsRDD(@(transient @param) sc: SparkContext,
```
what's this construction?

cseed replied:
https://stackoverflow.com/questions/3769624/annotating-constructor-parameters-in-scala

Oops, this should actually be `@(transient @field)`.  Will fix.

tpoterba commented on src/main/scala/is/hail/variant/Genotype.scala:
```diff
@@ -139,94 +139,137 @@ abstract class Genotype extends Serializable {
+      val dpq = queryInt("DP")
+      val gqq = queryInt("GQ")
+
+      if (s.hasField("PL") || !s.hasField("GT")) {
```
I'm uncomfortable with this logic. I realize that this was already in Hail in some form, but it makes me squirm a bit.

tpoterba commented on src/main/scala/is/hail/annotations/MemoryBlock.scala:
```diff
@@ -187,19 +248,367 @@ final class MemoryBuffer(sizeHint: Int = 128) {
   }
 }
+
+case class RegionValue(region: MemoryBuffer, offset: Int)
```
Wait... I know you... `Pointer`!

tpoterba commented on src/main/scala/is/hail/annotations/MemoryBlock.scala:
```diff
@@ -187,19 +248,367 @@ final class MemoryBuffer(sizeHint: Int = 128) {
+class RegionValueBuilder(region: MemoryBuffer) {
+  var start: Int = _
+
+  val typestk = new mutable.Stack[Type]()
```
These are all implemented in terms of lists, which is probably part of the reason this is slow.

cseed replied:
I know.  Odersky should be ashamed.  Profiling didn't turn this up to be a hot spot (yet), so I left it for now.

tpoterba commented on src/main/scala/is/hail/annotations/MemoryBlock.scala:
```diff
@@ -187,19 +248,367 @@ final class MemoryBuffer(sizeHint: Int = 128) {
+  val offsetstk = new mutable.Stack[Int]()
+  val elementsOffsetstk = new mutable.Stack[Int]()
+
+  def current(): (Type, Int) = {
```
egregious tuple allocation demerit

tpoterba commented on src/main/scala/is/hail/io/vcf/LoadVCF.scala:
```diff
@@ -103,24 +100,60 @@ object LoadVCF {
-        .map { case (line, i) => headerField(line, i, callFields) }
-        .toArray))
-    else None
+  def headerSignatue[T <: VCFCompoundHeaderLine](lines: java.util.Collection[T],
```
Typo

cseed commented on src/main/scala/is/hail/annotations/UnsafeRow.scala:
```diff
@@ -1,134 +1,291 @@
-      val bitShift = i & 0x7
-      val missingByte = mb.loadByte(start + 4 + byteIndex)
-      val isMissing = (missingByte & (0x1 << bitShift)) != 0
+  def readArrayAltAllele(region: MemoryBuffer, offset: Int): Array[AltAllele] = {
```
I can't because I don't have the right BroadcastTypeTree.  I originally had BroadcastTypeTree contain a broadcast of all the complex standard types, but that was a huge mess.  I think this is cleaner.  (And I also temporary.)

tpoterba replied:
got it.

cseed commented on src/main/scala/is/hail/expr/Type.scala:
```diff
@@ -190,6 +187,10 @@ sealed abstract class Type extends Serializable { self =>
+
+  /*  Fundamental types are types that can be handled natively by RegionValueBuilder: primitive
+      types, Array and Struct. */
+  def fundamentalType: Type = this
```
I did?  Oops.  How were you using it?

We still want them for resolving functions in the function registry, right?  Also, TVariant will soon contain the reference.

tpoterba replied:
I think this is the right thing - each type contains its representation in primitives. 

I don't mean remove TVariant and TAltAllele, I mean remove the ComplexType class, since `def representation`is just `fundamentalType` for types that actually have a different fundamental type than `this`

cseed commented on src/main/scala/is/hail/annotations/MemoryBlock.scala:
```diff
@@ -76,6 +88,11 @@ final class MemoryBlock(val mem: Array[Long]) {
     Platform.copyMemory(bytes, Platform.BYTE_ARRAY_OFFSET, mem, Platform.LONG_ARRAY_OFFSET + off, bytes.length)
   }
 
+  def storeBytes(off: Int, bytes: Array[Byte], size: Int) {
```
I think this is actually "consistent".  size modifies bytes (the source) here.  In `loadBytes(off, size, bytes)`, n modifies `this` (the source), therefore it goes with off, before the destination.

cseed commented:
@tpoterba Back to you.

Addressed comments.  Nuked MemoryBlock, moved the array to MemoryBuffer.  I think I made the safety tests in MemoryBuffer complete.  I changed the array to Array[Byte].  It is working.  There might still be an alignment issue (x86 supports unaligned loads but with possible performance penalty) but I'm OK leaving it to be addressed separately.

I think ComplexType is good but I agree we can remove representation and just fundamentalType.

cseed commented:
Nuked `UnsafeRowBuilder` and associated tests.  There was a bug in the tests that the stronger asserts in MemoryBuffer caught.  Easier to nuke than fix, since we aren't planning to use it.
